x-pack/plugins/security/public/session/session_timeout.ts

```diff
@@ -0,0 +1,309 @@
+      this.stopVisibilityMonitor = this.stopVisibilityMonitor?.();
+    }
+  };
+
+  private shouldExtend() {
+    const { lastExtensionTime } = this.sessionState$.getValue();
+    return (
+      !this.isFetchingSessionInfo &&
+      !this.warningToast &&
+      Date.now() > lastExtensionTime + SESSION_EXTENSION_THROTTLE_MS
+    );
+  }
+
+  private fetchSessionInfo = async (extend = false) => {
+    this.isFetchingSessionInfo = true;
+    try {
+      const sessionInfo = await this.http.fetch<SessionInfo | ''>(SESSION_ROUTE, {
+        method: extend ? 'POST' : 'GET',
+        asSystemRequest: !extend,
+      });
+      if (sessionInfo) {
+        const { expiresInMs, canBeExtended } = sessionInfo;
+        const nextState = {
+          lastExtensionTime: Date.now(),
+          expiresInMs,
+          canBeExtended,
+        };
+        this.sessionState$.next(nextState);
+        if (this.channel) {
+          this.channel.postMessage(nextState);
+        }
+      }
+    } catch (error) {
+      // ignore
+    } finally {
+      this.isFetchingSessionInfo = false;
+    }
+  };
+
+  private showWarning = () => {
+    if (!this.warningToast) {
+      const onExtend = () => this.fetchSessionInfo(true);
+      const onClose = () => (this.warningToast = undefined);
+      const toast = createSessionExpirationToast(this.sessionState$, onExtend, onClose);
+      this.warningToast = this.notifications.toasts.add(toast);
+    }
+  };
+
+  private hideWarning = () => {
+    if (this.warningToast) {
+      this.notifications.toasts.remove(this.warningToast);
+      this.warningToast = undefined;
+    }
+  };
+}
+
+/**
+ * Starts a timer that uses a native `setTimeout` under the hood. When `timeout` is larger
+ * than the maximum supported one then method calls itself recursively as many times as needed.
+ * @param callback A function to be executed after the timer expires.
+ * @param timeout The time, in milliseconds the timer should wait before the specified function is
+ * executed.
+ * @returns Function to stop the timer.
+ */
+export function startTimer(
+  callback: () => void,
+  timeout: number,
+  updater?: (id: NodeJS.Timeout) => void
+) {
+  // Max timeout is the largest possible 32-bit signed integer or 2,147,483,647 or 0x7fffffff.
+  const maxTimeout = 0x7fffffff;
+  let timeoutID: NodeJS.Timeout;
+  updater = updater ?? ((id: NodeJS.Timeout) => (timeoutID = id));
+
+  updater(
+    timeout > maxTimeout
+      ? setTimeout(() => startTimer(callback, timeout - maxTimeout, updater), maxTimeout)
+      : setTimeout(callback, timeout)
+  );
+
+  return () => clearTimeout(timeoutID);
+}
+
+/**
+ * Adds event handlers to the window object that track user activity.
+ * @param callback Function to be executed when user activity is detected.
+ * @returns Function to remove all event handlers from window.
+ */
+function monitorActivity(callback: () => void) {
+  const eventTypes = ['mousemove', 'mousedown', 'wheel', 'touchstart', 'keydown'];
```

Comment:
Thinking about this from an a11y perspective: I am guessing that because we included `keydown` here, using a screen reader will also count as session activity :+1:.

I imagine that screen readers will also keep the window visible so user activity will still be captured. I'm not entirely sure, and I've never used a screen reader before. If you haven't already done so, it would be good to run this by someone knowledgeable in this area.